### PR TITLE
Using Django in Kivy applications

### LINKED ISSUE #3608: Using Django in Kivy applications.
We ask that you please be patient as our team works through approving and publishing all articles/tutorials in a timely manner. Allow 1-3 days for a topic to be reviewed and/or approved - allow 3-7 days for an articles to be reviewed and/or published.

Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles to go over.

**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**NOTE: We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

*We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.*

### Proposed title of article
Using Django in Kivy applications..

### Introduction paragraph (2-3 paragraphs):
Kivy is a python library used to help build different applications that can run on various platforms.You can work with buttons, labels , scrollview, canvas etc.We will use Kivy to build  a simple todo application and see how you could use django to manipulate data from a web server. 

### Key takeaways: 
1.Learn how to work with the Kivy library.
2. Improve your python skills.

### References: 
Please list links to any published content/research that you intend to use to support/guide this article.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by JohnKiguru**
### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #3608

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by Qodestackr** [comment]
Hi @JohnKiguru I'd like to be your test reviewer for this article :+1:

**Comment by srishilesh** [comment]
@JohnKiguru 

On a quick glance, I don't think this article is ready for review.

I see that there are 2 `index.md` files - `content/articles/using-django-in-kivy-framework/index.md` and `content/articles/DjangoKivy/index.md`. Please remove the one that's not needed.

I suggest you rename the article to more meaningful one.

And, the image URLs are not right.

I would suggest you to take a look at your previous article reviews, correct them as needed, and make a ready to publish PR.

Thanks.

**Comment by JohnKiguru** [comment]
I think I have done that...

**Comment by srishilesh** [comment]
@JohnKiguru 

Looks good.

There's one more change left.

Rename `content/articles/DjangoKivy/` -> `content/articles/using-django-in-kivy-application/` for all the files.

**Comment by JohnKiguru** [comment]
I think am done...

**Comment by srishilesh** [comment]
@JohnKiguru 

Perfect. This looks good.

Will start the review soon.

File: content/articles/using-django-in-kivy-application/index.md [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+    path('', include('tasks.urls'))
+]
+```
+We can now run `python3 manage.py makemigrations` and `python3 manage.py migrate`.
+
+When we run `python3 manage.py runserver`, we should have a page similar to this:
+
+`http://127.0.0.1:8000/`
+
+![Tasks Home Page](section-engineering-education/engineering-education/content/articles/using-djnago-in-kivy-application/tasks_home_page.png)
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:162]:
```suggestion
![Tasks Home Page](/engineering-education/content/articles/using-djnago-in-kivy-application/tasks_home_page.png)
```

File: content/articles/using-django-in-kivy-application/index.md [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+
+When you create a task instance as follows:
+
+`http://127.0.0.1:8000/create`
+
+![Create Task Page](section-engineering-education/engineering-education/content/articles/using-django-in-kivy-application/create_task_page.png)
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:168]:
```suggestion
![Create Task Page](/engineering-education/content/articles/using-django-in-kivy-application/create_task_page.png)
```

File: content/articles/using-django-in-kivy-application/index.md [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+
+You should be able to see something similar to this:
+
+`http://127.0.0.1:8000/`
+
+![Task Created](section-engineering-education/engineering-education/content/articles/using-django-in-kivy-application/task_created.png)
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:174]:
```suggestion
![Task Created](/engineering-education/content/articles/using-django-in-kivy-application/task_created.png)
```

File: content/articles/using-django-in-kivy-application/index.md [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+
+class TodoApp(App):
+    pass
+
+if __name__ == '__main__':
+    TodoApp().run()
+```
+When you run the file, you should get the following result:
+
+![Original Kivy App](section-engineering-education/engineering-education/content/articles/using-django-in-kivy-application/original_kivy_app.png)
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:195]:
```suggestion
![Original Kivy App](/engineering-education/content/articles/using-django-in-kivy-application/original_kivy_app.png)
```

File: content/articles/using-django-in-kivy-application/index.md [hunk 5]
```diff
@@ -0,0 +1,596 @@
...
+    the buttons we will use to explore the application.
+2. We then proceed by creating the `ScreenManger` class that manages the display of what is displayed where in the application.
+3. In the `todo.kv` file, we declare a `BoxLayout` as our main interface that our program will show.We then 
+    We give the menu a position of bottom set by ` size_hint_y: .1`. We also declare that it's managed by `ScreenManager` by declaring ` manager: screen_manager`.
+    By setting the `id` property of the `ScreenManager` as `screen_manager` the position of the Menu is now on top.
+4. We the declare the properties of our `Menu` class. The class will have an action bar which will contain two buttons, the `Home` action button and the `Add New` action button.
+
+Your application should be similar to the one below:
+
+![Creating The Action Bar](section-engineering-education/engineering-education/content/articles/using-django-in-kivy-application/creating_the_action_bar.png)
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:264]:
```suggestion
![Creating The Action Bar](/engineering-education/content/articles/using-django-in-kivy-application/creating_the_action_bar.png)
```

File: content/articles/using-django-in-kivy-application/index.md [hunk 6]
```diff
@@ -0,0 +1,596 @@
...
+The `MyRecycleView` class is initialized by having a function `load_data` that make requests to our server using the `requests` library.
+The data is appended to a list containing dictionaries of our tasks with the key `text`. The function returns the list as a `data` variable.
+The function is called every second by setting a clock interval of 1.
+
+In the `todo.kv` file, we replace the contents of our `HomeScreen` with a `BoxLayout` that contains our `MyRecycleView` class.
+We then declare the properties of our `MyRecycleView ` as having BoxLayout that contains a list of labels, each whose text is a task received from our server.
+
+You should now be able to see the `code` task we created in our web-based interface earlier.
+
+![Displaying Tasks from server](section-engineering-education/engineering-education/content/articles/using-django-in-kivy-application/displaying_tasks_from_server.png)
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:442]:
```suggestion
![Displaying Tasks from server](/engineering-education/content/articles/using-django-in-kivy-application/displaying_tasks_from_server.png)
```

File: content/articles/using-django-in-kivy-application/index.md [hunk 7]
```diff
@@ -0,0 +1,596 @@
...
+        text: 'Submit'
+        on_release: root.submit_input()
+```
+The `AddNewForm` contains a function `submit_input` that makes a post request to our server passing the input text as data.
+The form has a textinput that one can type the task, and a button that calls the `submit_input` function when released.
+We then declare a `BoxLayout ` class that will contain the `AddNewForm` and a label text "Add to list..."
+
+When you now click `Add New` you should now see the following:
+
+![Creating a task](section-engineering-education/engineering-education/content/articles/using-django-in-kivy-application/creating_a_task.png)
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:587]:
```suggestion
![Creating a task](/engineering-education/content/articles/using-django-in-kivy-application/creating_a_task.png)
```

File: content/articles/using-django-in-kivy-application/index.md [hunk 8]
```diff
@@ -0,0 +1,596 @@
...
+
+When you create a task, say "Complete application" and click submit button once and click `Home`, you should have the following:
+
+![Task submitted successfully](section-engineering-education/engineering-education/content/articles/using-django-in-kivy-application/task_submitted_successfully.png)
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:591]:
```suggestion
![Task submitted successfully](/engineering-education/content/articles/using-django-in-kivy-application/task_submitted_successfully.png)
```

File: content/articles/using-django-in-kivy-application/index.md [hunk 9]
```diff
@@ -0,0 +1,596 @@
...
+
+### Conclusion. 
+We have learnt how create a kivy application and how to use Django in such applications. You can therefore
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:594]: Conclusion could have been more elaborate.

Give a summary of what was covered.

Add a topic `Further reading` that lists out various reference links

File: content/articles/using-django-in-kivy-application/index.md [hunk 10]
```diff
@@ -0,0 +1,596 @@
+### Using Django in Kivy applications.
+
+### Introduction.
+
+Kivy is a popular library for developing applications that can run across all major platforms.
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:5]: `developing applications`

Be specific. Applications can be anything -front end, backend. What does Kivy help to build?

File: content/articles/using-django-in-kivy-application/index.md [hunk 11]
```diff
@@ -0,0 +1,596 @@
...
+There may, however, be situations that require your application to access and manipulate data being provided by a web server 
+hosted at a site.In this tutorial, we are going to learn how to make requests to a server
+using the `requests` library. We are going to use Django to develop the server that holds the data we want to manipulate.
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:8]: Here, every sentence is in a new line.

If possible, make them into a paragraph. One para must talk about one idea

File: content/articles/using-django-in-kivy-application/index.md [hunk 12]
```diff
@@ -0,0 +1,596 @@
...
+
+### Prerequisites.
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:10]: Remove `.`

Comment (srishilesh) [reply content/articles/using-django-in-kivy-application/index.md:10]: Do this for all the topic headings

File: content/articles/using-django-in-kivy-application/index.md [hunk 13]
```diff
@@ -0,0 +1,596 @@
...
+    └── TodoAPI
+        ├── asgi.py
+        ├── __init__.py
+        ├── settings.py
+        ├── urls.py
+        └── wsgi.py
+
+
+
+```
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:54]:
```suggestion
.
└── TodoAPI
    ├── manage.py
    ├── tasks
    │   ├── admin.py
    │   ├── apps.py
    │   ├── __init__.py
    │   ├── migrations
    │   │   └── __init__.py
    │   ├── models.py
    │   ├── tests.py
    │   └── views.py
    └── TodoAPI
        ├── asgi.py
        ├── __init__.py
        ├── settings.py
        ├── urls.py
        └── wsgi.py
```
```

File: content/articles/using-django-in-kivy-application/index.md [hunk 14]
```diff
@@ -0,0 +1,596 @@
...
+```python
+INSTALLED_APPS = [
+    'django.contrib.admin',
+    'django.contrib.auth',
+    'django.contrib.contenttypes',
+    'django.contrib.sessions',
+    'django.contrib.messages',
+    'django.contrib.staticfiles',
+    'tasks',
+    'rest_framework'
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:67]: What are these for? Please explain if anything is specific to the app you are building

File: content/articles/using-django-in-kivy-application/index.md [hunk 15]
```diff
@@ -0,0 +1,596 @@
...
+from django.db import models
+
+# Create your models here.
+class Task(models.Model):
+    name = models.CharField(max_length=200)
+    def __str__(self):
+        return self.name
+```
+
+Let's create a `serializer.py` file that handles serialization and deserialization of task instances.Add the following lines of code to it:
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:83]: `serialization and deserialization`

Explain in 1 - 2 lines about these, and reference links for these

File: content/articles/using-django-in-kivy-application/index.md [hunk 16]
```diff
@@ -0,0 +1,596 @@
...
+from .models import *
+
+
+class TaskSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Task
+        field = '__all__'
+
+```
+We are going to also create some views to render data to a web page. There are different ways to create the views and we will use
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:98]: Give 1 line explanation of what views are? And, what function-based views are?

File: content/articles/using-django-in-kivy-application/index.md [hunk 17]
```diff
@@ -0,0 +1,596 @@
...
+    return Response(serializer.data, status=status.HTTP_200_OK)
+
+
+@api_view(['POST'])
+def create_task(request):
+    data = request.data
+    serializer = TaskSerializer(data=data)
+    if serializer.is_valid():
+        return Response(serializer.data, status=status.HTTP_200_OK)
+
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:128]: Explain this.

File: content/articles/using-django-in-kivy-application/index.md [hunk 18]
```diff
@@ -0,0 +1,596 @@
...
+```
+We then create a `urls.py` file to for routing our views. Create the file and add the following code:
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:130]: Can we create a subtopic for this? Like `#### Routing views`.

File: content/articles/using-django-in-kivy-application/index.md (hunk 16 above)
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:98]: Subtopic `#### Creating views`

File: content/articles/using-django-in-kivy-application/index.md [hunk 19]
```diff
@@ -0,0 +1,596 @@
...
+`tasks/urls.py`
+
+```python
+from django.urls import path
+from . import views
+
+urlpatterns = [
+    path('', views.all_tasks, name='all_tasks'),
+    path('create', views.create_task, name='create_task')
+]
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:141]: Explain this mapping of routes with functions

File: content/articles/using-django-in-kivy-application/index.md [hunk 20]
```diff
@@ -0,0 +1,596 @@
...
+`TodoAPI/urls.py`
+
+```python
+from django.contrib import admin
+from django.urls import path, include
+
+urlpatterns = [
+    path('admin/', admin.site.urls),
+    path('', include('tasks.urls'))
+]
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:154]: Explain

File: content/articles/using-django-in-kivy-application/index.md [hunk 21]
```diff
@@ -0,0 +1,596 @@
...
+`http://127.0.0.1:8000/`
+
+![Task Created](section-engineering-education/engineering-education/content/articles/using-django-in-kivy-application/task_created.png)
+
+Our Django API is working and we can now proceed to creating the `kivy` application.
+
+### Creating the Kivy application.
+Let's create a `main.py` in a folder of choice.I am going to keep the file in the same directory as our `TodoAPI` project folder.
+We are going to get started with these lines of code in our `main.py` file.The `TodoApp` is our main entrypoint
+of our application and every execution begins from there.
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:181]: Make it into a paragraph.

File: content/articles/using-django-in-kivy-application/index.md [hunk 22]
```diff
@@ -0,0 +1,596 @@
...
+class ScreenManagement(ScreenManager):
+    pass
+#app class
+class TodoApp(App):
+    pass
+  
+
+if __name__ == '__main__':
+    TodoApp().run()
+```
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:221]: ` The description of what is happening will be explained after the code`

???

File: content/articles/using-django-in-kivy-application/index.md [hunk 23]
```diff
@@ -0,0 +1,596 @@
...
+        text: "Home"
+<AddScreen>:
+    Label:
+        text: "Add to list..."
+
+```
+We declare the `HomeScreen` class and give it a `Label` instance with the text "home".We also do the same for the `AddScreen` with the label text "Add to list..."
+We have given both screens as children of `ScreenManagement`. We also have given functionality to our action buttons by 
+making them being able to access the screen by defining ` on_press: root.manager.current = 'screen_home'` and ` on_press: root.manager.current = 'screen_add'`. You should now be able to explore the two screens 
+and see the "Home" text in the `HomeScreen` and "Add to list..." in the `AddScreen`
```
Comment (srishilesh) [inline content/articles/using-django-in-kivy-application/index.md:346]: Make it as bulleted points

**Review by srishilesh — CHANGES_REQUESTED** [review]
@JohnKiguru 

Great article.

I've suggested a few changes. Please let me know when done.

Make sure to perform a grammar check.

Thanks.

**Comment by hectorkambow** [comment]
@JohnKiguru 
Good afternoon and thank you for submitting your topic to the EngEd program. 
If it seems this PR is still quite a bit aways from being completed. 
As a rule of thumb - please only submit articles (pull requests) with a complete and fully polished article - ready to be published. 
All revisions and edits should be completed on your own forked repo (as to not create a huge backlog in the queue). 
We will be closing this and please re-submit when it is ready to be published as a new PR (and you can add the duplicate label) and ping the reviewer and myself.

**Comment by srishilesh** [comment]
@hectorkambow 

Reopening this PR , since the author has made requested changes.

Will close again if there's anymore changes left.

**Comment by ahmadmardeni1** [comment]
plagiarism check done ✅

**Comment by srishilesh** [comment]
I'll complete my review by tomorrow

**Comment by JohnKiguru** [comment]
Great...

**Comment by srishilesh** [comment]
I'll complete the review by today. Apologies for the delay

**Review by srishilesh — APPROVED** [review]
@JohnKiguru 

Thanks for making the changes.

It looks great now. The tutorial was very beginner friendly. Though, the code snippets were understandable, it could have been explained more briefly.

Please make sure to go through my commits.

Moving for the final review.

Thanks.

**Comment by JohnKiguru** [comment]
No worries...

**Comment by Kayere** [comment]
Great article @JohnKiguru.
Thank you for the review @srishilesh.

**Pull request merged**